data/model/oci/manifest.py

```diff
@@ -392,6 +321,131 @@ def _create_manifest(
     return CreatedManifest(manifest=manifest, newly_created=True, labels_to_apply=labels_to_apply)
 
 
+def _build_blob_map(
+    repository_id,
+    manifest_interface_instance,
+    retriever,
+    storage,
+    raise_on_error=False,
+    require_empty_layer=True,
+):
+    """ Builds a map containing the digest of each blob referenced by the given manifest,
+        to its associated Blob row in the database. This method also verifies that the blob
+        is accessible under the given repository. Returns None on error (unless raise_on_error
+        is specified). If require_empty_layer is set to True, the method will check if the manifest
+        references the special shared empty layer blob and, if so, add it to the map. Otherwise,
+        the empty layer blob is only returned if it was *explicitly* referenced in the manifest.
+        This is necessary because Docker V2_2/OCI manifests can implicitly reference an empty blob
+        layer for image layers that only change metadata.
+    """
+
+    # Ensure all the blobs in the manifest exist.
+    digests = set(manifest_interface_instance.local_blob_digests)
+    blob_map = {}
+
+    # If the special empty layer is required, simply load it directly. This is much faster
+    # than trying to load it on a per repository basis, and that is unnecessary anyway since
+    # this layer is predefined.
+    if EMPTY_LAYER_BLOB_DIGEST in digests:
+        digests.remove(EMPTY_LAYER_BLOB_DIGEST)
+        blob_map[EMPTY_LAYER_BLOB_DIGEST] = get_shared_blob(EMPTY_LAYER_BLOB_DIGEST)
+        if not blob_map[EMPTY_LAYER_BLOB_DIGEST]:
+            if raise_on_error:
+                raise CreateManifestException("Unable to retrieve specialized empty blob")
+
+            logger.warning("Could not find the special empty blob in storage")
```

Comment:
I am not sure under which conditions this would happen. Should the error output be a bit more actionable? Something along the lines of:

```
raise CreateManifestException("Unable to retrieve specialized empty blob. Has any data been pushed?")
```
I don't actually know what the correct action or steps to take would be. I'm just thinking that this warning is useful but I don't know why it's occurring or how to fix it. I imagine _if_ it happened, it'd be an important thing to solve as it's used quite often.

It might also help in the "self-sevice" direction.

Reply:
It shouldn't really ever happen, since we no longer actually use the empty layer in normal pushes. If it does, then its an internal error, and it should be reported